// eslint-plugin-react-debug/dist/index.js
//#region rolldown:runtime
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (all) => {
	let target = {};
	for (var name$2 in all) __defProp(target, name$2, {
		get: all[name$2],
		enumerable: true
	});
	return target;
};
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
let __eslint_react_shared = require("@eslint-react/shared");
__eslint_react_shared = __toESM(__eslint_react_shared);
let __eslint_react_core = require("@eslint-react/core");
__eslint_react_core = __toESM(__eslint_react_core);
let __typescript_eslint_utils = require("@typescript-eslint/utils");
__typescript_eslint_utils = __toESM(__typescript_eslint_utils);
let __eslint_react_eff = require("@eslint-react/eff");
__eslint_react_eff = __toESM(__eslint_react_eff);
let __eslint_react_kit = require("@eslint-react/kit");
__eslint_react_kit = __toESM(__eslint_react_kit);
let __typescript_eslint_types = require("@typescript-eslint/types");
__typescript_eslint_types = __toESM(__typescript_eslint_types);
let ts_pattern = require("ts-pattern");
ts_pattern = __toESM(ts_pattern);

//#region src/configs/all.ts
var all_exports = __export({
	name: () => name$1,
	rules: () => rules,
	settings: () => settings
});
const name$1 = "react-debug/all";
const rules = {
	"react-debug/class-component": "warn",
	"react-debug/function-component": "warn",
	"react-debug/hook": "warn",
	"react-debug/is-from-react": "warn",
	"react-debug/jsx": "warn"
};
const settings = { "react-x": __eslint_react_shared.DEFAULT_ESLINT_REACT_SETTINGS };

//#endregion
//#region package.json
var name = "eslint-plugin-react-debug";
var version = "1.53.1";

//#endregion
//#region src/utils/create-rule.ts
const createRule = __typescript_eslint_utils.ESLintUtils.RuleCreator((0, __eslint_react_shared.getDocsUrl)("debug"));

//#endregion
//#region src/utils/stringify.ts
function stringify(value) {
	return JSON.stringify(value, null, 2);
}

//#endregion
//#region src/rules/class-component.ts
const RULE_NAME$4 = "class-component";
const RULE_FEATURES$4 = ["DBG"];
var class_component_default = createRule({
	meta: {
		type: "problem",
		docs: {
			description: "Reports all class components.",
			[Symbol.for("rule_features")]: RULE_FEATURES$4
		},
		messages: { classComponent: "{{json}}" },
		schema: []
	},
	name: RULE_NAME$4,
	create: create$4,
	defaultOptions: []
});
function create$4(context) {
	const { ctx, listeners } = __eslint_react_core.useComponentCollectorLegacy();
	return {
		...listeners,
		"Program:exit"(program) {
			const components = ctx.getAllComponents(program);
			for (const { name: name$2 = "anonymous", node: component } of components.values()) context.report({
				messageId: "classComponent",
				node: component,
				data: { json: stringify({ name: name$2 }) }
			});
		}
	};
}

//#endregion
//#region src/rules/function-component.ts
const RULE_NAME$3 = "function-component";
const RULE_FEATURES$3 = ["DBG"];
var function_component_default = createRule({
	meta: {
		type: "problem",
		docs: {
			description: "Reports all function components.",
			[Symbol.for("rule_features")]: RULE_FEATURES$3
		},
		messages: { functionComponent: "{{json}}" },
		schema: []
	},
	name: RULE_NAME$3,
	create: create$3,
	defaultOptions: []
});
function create$3(context) {
	const { ctx, listeners } = __eslint_react_core.useComponentCollector(context, {
		collectDisplayName: true,
		collectHookCalls: true,
		hint: __eslint_react_core.DEFAULT_COMPONENT_DETECTION_HINT
	});
	return {
		...listeners,
		"Program:exit"(program) {
			const components = ctx.getAllComponents(program);
			for (const { name: name$2 = "anonymous", node, displayName, flag, hookCalls } of components.values()) context.report({
				messageId: "functionComponent",
				node,
				data: { json: stringify({
					name: name$2,
					displayName: displayName == null ? "none" : context.sourceCode.getText(displayName),
					forwardRef: (flag & __eslint_react_core.ComponentFlag.ForwardRef) > 0n,
					hookCalls: hookCalls.length,
					memo: (flag & __eslint_react_core.ComponentFlag.Memo) > 0n
				}) }
			});
		}
	};
}

//#endregion
//#region src/rules/hook.ts
const RULE_NAME$2 = "hook";
const RULE_FEATURES$2 = ["DBG"];
var hook_default = createRule({
	meta: {
		type: "problem",
		docs: {
			description: "Reports all React Hooks.",
			[Symbol.for("rule_features")]: RULE_FEATURES$2
		},
		messages: { hook: "{{json}}" },
		schema: []
	},
	name: RULE_NAME$2,
	create: create$2,
	defaultOptions: []
});
function create$2(context) {
	const { ctx, listeners } = __eslint_react_core.useHookCollector();
	return {
		...listeners,
		"Program:exit"(program) {
			const allHooks = ctx.getAllHooks(program);
			for (const { name: name$2, node, hookCalls } of allHooks.values()) context.report({
				messageId: "hook",
				node,
				data: { json: stringify({
					name: name$2,
					hookCalls: hookCalls.length
				}) }
			});
		}
	};
}

//#endregion
//#region src/rules/is-from-react.ts
const RULE_NAME$1 = "is-from-react";
const RULE_FEATURES$1 = ["DBG"];
var is_from_react_default = createRule({
	meta: {
		type: "problem",
		docs: {
			description: "Reports all identifiers that are initialized from React.",
			[Symbol.for("rule_features")]: RULE_FEATURES$1
		},
		messages: { isFromReact: "{{json}}" },
		schema: []
	},
	name: RULE_NAME$1,
	create: create$1,
	defaultOptions: []
});
function create$1(context) {
	const { importSource = "react" } = (0, __eslint_react_shared.getSettingsFromContext)(context);
	function visitorFunction(node) {
		if (node.parent.type === __typescript_eslint_utils.AST_NODE_TYPES.ImportSpecifier && node.parent.imported === node && node.parent.imported.name === node.parent.local.name) return;
		const name$2 = node.name;
		const initialScope = context.sourceCode.getScope(node);
		if (!isFromReact(node, importSource, initialScope)) return;
		context.report({
			messageId: "isFromReact",
			node,
			data: { json: stringify({
				name: name$2,
				importSource
			}) }
		});
	}
	return {
		Identifier: visitorFunction,
		JSXIdentifier: visitorFunction
	};
}
/**
* Check if an identifier node is initialized from React
* @param node The identifier node to check
* @param importSource The import source to check against
* @param initialScope Initial scope to search for the identifier
* @returns Whether the identifier node is initialized from React
*/
function isFromReact(node, importSource, initialScope) {
	const name$2 = node.name;
	switch (true) {
		case node.parent.type === __typescript_eslint_utils.AST_NODE_TYPES.MemberExpression && node.parent.property === node && node.parent.object.type === __typescript_eslint_utils.AST_NODE_TYPES.Identifier: return __eslint_react_core.isInitializedFromReact(node.parent.object.name, importSource, initialScope);
		case node.parent.type === __typescript_eslint_utils.AST_NODE_TYPES.JSXMemberExpression && node.parent.property === node && node.parent.object.type === __typescript_eslint_utils.AST_NODE_TYPES.JSXIdentifier: return __eslint_react_core.isInitializedFromReact(node.parent.object.name, importSource, initialScope);
		default: return __eslint_react_core.isInitializedFromReact(name$2, importSource, initialScope);
	}
}

//#endregion
//#region src/rules/jsx.ts
const { JsxEmit } = __eslint_react_kit.JsxConfig;
const RULE_NAME = "jsx";
const RULE_FEATURES = ["DBG"];
var jsx_default = createRule({
	meta: {
		type: "problem",
		docs: {
			description: "Reports all JSX elements and fragments.",
			[Symbol.for("rule_features")]: RULE_FEATURES
		},
		messages: { jsx: "{{json}}" },
		schema: []
	},
	name: RULE_NAME,
	create,
	defaultOptions: []
});
function create(context) {
	const jsxConfigFromContext = __eslint_react_kit.JsxConfig.getFromContext(context);
	const jsxConfigFromAnnotation = __eslint_react_kit.JsxConfig.getFromAnnotation(context);
	const jsxConfig = {
		...jsxConfigFromContext,
		...jsxConfigFromAnnotation
	};
	function getReportDescriptor(context$1) {
		return (node) => ({
			messageId: "jsx",
			node,
			data: { json: stringify({
				kind: (0, ts_pattern.match)(node).with({ type: __typescript_eslint_types.AST_NODE_TYPES.JSXElement }, (n) => __eslint_react_core.isFragmentElement(context$1, n) ? "fragment" : "element").with({ type: __typescript_eslint_types.AST_NODE_TYPES.JSXFragment }, () => "fragment").exhaustive(),
				type: __eslint_react_core.getElementType(context$1, node),
				jsx: (0, ts_pattern.match)(jsxConfig.jsx).with(JsxEmit.None, () => "none").with(JsxEmit.ReactJSX, () => "react-jsx").with(JsxEmit.ReactJSXDev, () => "react-jsx-dev").with(JsxEmit.React, () => "react").with(JsxEmit.ReactNative, () => "react-native").with(JsxEmit.Preserve, () => "preserve").otherwise(() => "unknown"),
				jsxFactory: jsxConfig.jsxFactory,
				jsxFragmentFactory: jsxConfig.jsxFragmentFactory,
				jsxImportSource: jsxConfig.jsxImportSource,
				jsxRuntime: (0, ts_pattern.match)(jsxConfig.jsx).with(ts_pattern.P.union(JsxEmit.None, JsxEmit.ReactJSX, JsxEmit.ReactJSXDev), () => "automatic").otherwise(() => "classic")
			}) }
		});
	}
	return { "JSXElement, JSXFragment": (0, __eslint_react_eff.flow)(getReportDescriptor(context), __eslint_react_kit.Reporter.make(context).send) };
}

//#endregion
//#region src/plugin.ts
const plugin = {
	meta: {
		name,
		version
	},
	rules: {
		["class-component"]: class_component_default,
		["function-component"]: function_component_default,
		["hook"]: hook_default,
		["is-from-react"]: is_from_react_default,
		["jsx"]: jsx_default,
		"react-hooks": hook_default
	}
};

//#endregion
//#region src/index.ts
const { toFlatConfig, toLegacyConfig } = (0, __eslint_react_shared.getConfigAdapters)("react-debug", plugin);
var src_default = {
	...plugin,
	configs: {
		["all"]: toFlatConfig(all_exports),
		["all-legacy"]: toLegacyConfig(all_exports)
	}
};

//#endregion
module.exports = src_default;
